statement.go

```diff
@@ -148,56 +149,115 @@ func (s *stmt) start(args []driver.Value) error {
 	return nil
 }
 
+// Deprecated: Use ExecContext instead.
 func (s *stmt) Exec(args []driver.Value) (driver.Result, error) {
-	if s.closed {
-		panic("database/sql/driver: misuse of duckdb driver: Exec after Close")
-	}
-	if s.rows {
-		panic("database/sql/driver: misuse of duckdb driver: Exec with active Rows")
-	}
+	return s.ExecContext(context.Background(), argsToNamedArgs(args))
+}
 
-	err := s.start(args)
+func (s *stmt) ExecContext(ctx context.Context, nargs []driver.NamedValue) (driver.Result, error) {
+	res, err := s.executeWithCancellation(ctx, nargs)
 	if err != nil {
 		return nil, err
 	}
+	defer C.duckdb_destroy_result(res)
 
-	var res C.duckdb_result
-	if state := C.duckdb_execute_prepared(*s.stmt, &res); state == C.DuckDBError {
-		dbErr := C.GoString(C.duckdb_result_error(&res))
-		C.duckdb_destroy_result(&res)
-		C.duckdb_destroy_prepare(s.stmt)
-		return nil, errors.New(dbErr)
-	}
-	defer C.duckdb_destroy_result(&res)
-
-	ra := int64(C.duckdb_value_int64(&res, 0, 0))
-
+	ra := int64(C.duckdb_value_int64(res, 0, 0))
 	return &result{ra}, nil
 }
 
+// Deprecated: Use QueryContext instead.
 func (s *stmt) Query(args []driver.Value) (driver.Rows, error) {
+	return s.QueryContext(context.Background(), argsToNamedArgs(args))
+}
+
+func (s *stmt) QueryContext(ctx context.Context, nargs []driver.NamedValue) (driver.Rows, error) {
+	res, err := s.executeWithCancellation(ctx, nargs)
+	if err != nil {
+		return nil, err
+	}
+	s.rows = true
+	return newRowsWithStmt(*res, s), nil
+}
+
+// This method executes the query in steps and checks if context is cancelled before executing each step.
+// It uses Pending Result Interface C APIs to achieve this. Reference - https://duckdb.org/docs/api/c/api#pending-result-interface
+func (s *stmt) executeWithCancellation(ctx context.Context, nargs []driver.NamedValue) (*C.duckdb_result, error) {
```

Comment:
I would just call this `execute` or `query`. Cancellation is implied when accepting a ctx.